sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ChecksumCheckpointFileManager.scala

```diff
@@ -0,0 +1,513 @@
+    if (!verified) {
+      // It is possible the file was not read till the end by the reader,
+      // but we need the entire file content for checksum verification.
+      // Hence, we will read the file till the end from where reader stopped.
+      var remainingBytesRead = 0
+      val buffer = new Array[Byte](1024)
+      var bytesRead = 0
+      while ({ bytesRead = super.read(buffer); bytesRead != -1 }) {
+        remainingBytesRead += bytesRead
+      }
+
+      // we are at the end position so that tells us the size
+      val computedFileSize = mainStream.getPos
+
+      if (remainingBytesRead > 0) {
+        // Making this debug log since most of our files are not read exactly to the end
+        // and don't want to cause unnecessary noise in the logs.
+        logDebug(log"File ${MDC(PATH, path)} was not read till the end by reader. " +
+          log"Finished reading the rest of the file for checksum verification. " +
+          log"Remaining bytes read: ${MDC(NUM_BYTES, remainingBytesRead)}, " +
+          log"total size: ${MDC(NUM_BYTES, computedFileSize)}.")
+      }
+
+      // Read the expected checksum from the checksum file
+      val expectedChecksumJson = Source.fromInputStream(
+        expectedChecksumStream, StandardCharsets.UTF_8.name()).mkString
+      val expectedChecksum = Checksum.fromJson(expectedChecksumJson)
+      // Get what we computed while the main file was being read locally
+      // `in` is the CheckedSequentialInputStream we created
+      val computedChecksumValue = in.asInstanceOf[CheckedInputStream].getChecksum.getValue.toInt
+
+      logInfo(log"Verifying checksum for file ${MDC(PATH, path)}, " +
+        log"remainingBytesRead= ${MDC(NUM_BYTES, remainingBytesRead)}. " +
+        log"Computed(checksum= ${MDC(CHECKSUM, computedChecksumValue)}, " +
+        log"fileSize= ${MDC(NUM_BYTES, computedFileSize)})." +
+        log"Checksum file content: ${MDC(CHECKSUM, expectedChecksumJson)}")
+
+      verified = true
+
+      // Compare file size too, in case of collision
+      if (expectedChecksum.value.toInt != computedChecksumValue ||
+        expectedChecksum.mainFileSize != computedFileSize) {
+        throw QueryExecutionErrors.checkpointFileChecksumVerificationFailed(
+          path,
+          expectedSize = expectedChecksum.mainFileSize,
+          expectedChecksum.value,
+          computedSize = computedFileSize,
+          computedChecksumValue.toString)
+      }
+    }
+  }
+}
+
+/** This implements [[CheckedInputStream]] that allows us to compute the checksum as the file
+ * is being read. [[FSDataInputStream]] needs the passed in input stream to implement Seekable
+ * and PositionedReadable. We want the file to be read sequentially only to allow us to correctly
+ * compute the checksum. This is blocking the seekable apis from being used in the underlying stream
+ * */
+private class CheckedSequentialInputStream(data: InputStream)
+  // TODO: Make the checksum algo configurable
```

Comment:
lets add it in this change itself ? if you want to do this as a follow up - lets create the SPARK JIRA for this and link it here ?

Reply:
ticket created: https://issues.apache.org/jira/browse/SPARK-52009